Makefiles: add support to generate both `.hex` and `.bin` file and add FLASHFILE variable

cladmi opened the pull request:
### Contribution description

I need to be able to generate a `.bin` file to flash nucleo boards by drag-and-dropping the firmware.
The main goal of this PR main is to add a target for the bin file. This would allow asking for building the bin file from the command line but not creating it for every build.

* Remove the `HEXFILE = $(ELFFILE:.elf=.bin)` hack
* Add a `FLASHFILE` variable which fixes the hack in https://github.com/RIOT-OS/RIOT/pull/8801
* Change flasher script that where getting the flashfile from environment variable.

#### Integrated in sub PRs:

While looking into this I tried fixing other problems to not propagate hack somewhere else:
* [x] Put `ELFFILE` in its own target and generally break link in sub file targets
    * [x] ~~https://github.com/RIOT-OS/RIOT/pull/8844~~ separate target
    * [x] ~~https://github.com/RIOT-OS/RIOT/pull/8845~~ .hex and .bin targets
* [x] Fix boards generating a binary `.hex` file  ~~https://github.com/RIOT-OS/RIOT/pull/8840~~
* [x] Random refactoring ~~https://github.com/RIOT-OS/RIOT/pull/8839~~
* [x] Split most of the boards changes to take flashfile from command line arguments
  * [x] Use `BINFILE` instead of `HEXFILE` https://github.com/RIOT-OS/RIOT/pull/10471
    * [x] Flasher not even working for `common/remote` with PROGRAMMER=jlink https://github.com/RIOT-OS/RIOT/pull/10472
  * [x] board/*/flash.sh get flashfile from command line https://github.com/RIOT-OS/RIOT/pull/10473
  * [x] board/opencm904 https://github.com/RIOT-OS/RIOT/pull/10474
  * [x] jlink: get flash and debug files from cli https://github.com/RIOT-OS/RIOT/pull/10528
  * [x] openocd: get flash and debug files from cli https://github.com/RIOT-OS/RIOT/pull/10539
    * [x]  `hifive1` included here
  * [x] Other splits incoming out of https://github.com/RIOT-OS/RIOT/pull/9514
  * [x] ESP boards
  * [x] Other missing boards not currently migrated (see along the way)
* [x] Introduce FLASHFILE variable even if not used everywhere #11084
  * [x] Using `FLASHFILE`
    * [x] makefiles/murdock.inc.mk: do not overwrite FLASHFILE if set #11097
    * [x] tests/riotboot: use FLASHFILE for the generated file https://github.com/RIOT-OS/RIOT/pull/11089
    * [x] makefiles/mcuboot: prepare for when FLASHFILE is used #11112        
  * [x] Migrate boards by group
    * [x] makefiles/tools/jlink.inc.mk: use FLASHFILE #11130
    * [x] makefiles/edbg.inc.mk: use FLASHFILE #11172 
      * [x] makefiles/boot/riotboot.mk: fix unrelated introduced change #11284
    * [x] boards/lpc2k_pgm: use FLASHFILE for boards using lpc2k_pgm #11188
    * [x] tools/avrdude.inc.mk: use FLASHFILE variable #11190
    * [x] makefiles/tools/uniflash.inc.mk: add a common configuration for using UniFlash https://github.com/RIOT-OS/RIOT/pull/11198
    * [x] makefiles/openocd.inc.mk: use FLASHFILE #11254
    * [x] boards/cc2538: use FLASHFILE for boards using cc2538-bsl.py #11569
    * [x] boards/lobaro-lorabox: use FLASHFILE for boards using stm32loader.py #11629 
    * [x] cpu/esp*: use FLASHFILE for esp32 and esp8266 boards #11648
    * [x] makefiles/pyocd.inc.mk use FLASHFILE #11696 
    * [x] boards/pic32-xx: set FLASHFILE to HEXFILE #11699
    * [x] boards/mips-malta: set FLASHFILE #11711
    * [x] boards/mspdebug: use FLASHFILE for boards using mspdebug #11708
    * [x] makefiles/bossa.inc.mk: use FLASHFILE #11700
    * [x] boards/dfu-util: use FLASHFILE for boards using dfu-util #11709
    * [x] boards/single: use FLASHFILE #11710
    * [x] boards/native: define FLASHFILE #11752 
* [x] Cleanups
  * [x] makefiles/murdock.inc.mk: remove flashfile hack #11755
  * [x] Makefile.include: do not build HEXFILE by default anymore #11757 
  * [x]  makefiles: documentation for hexfile binfile #11845 
* [x] add a static test to verify it's done for all
* Other cleanups
  * [ ] Migrate `common/remote` and `cc2538dk` to use `jlink.inc.mk` similar to https://github.com/RIOT-OS/RIOT/pull/9851
  * [ ] tests/mcuboot: use FLASHFILE for the generated file (Waiting For Other PR #10494)
  * [ ] boards/hamilton: remove OFLAGS '-O binary' #11849
  * [ ] Remove overwriting `HEXFILE` in `docker`, it is now only handled to deduce it from `ELFFILE`
 

It has the side effect to not generate the '.hex' or '.bin' file if not required for the flasher, which I think is better

This ended up in a big PR that I will split.


#### Verifying after rebase:

Things I should look into after rebasing the PR in case new boards where added:

    git grep FFLAGS | grep -v FLASHFILE
    git grep OFLAGS |  grep -e '-O'
    git grep HEXFILE | grep -v FLASHFILE

### Issues/PRs references

* Its re-introducing https://github.com/RIOT-OS/RIOT/pull/1098 (reverted by https://github.com/RIOT-OS/RIOT/pull/1198)
* Will remove the hack for https://github.com/RIOT-OS/RIOT/pull/8801/files
* ~~It will need rebase after https://github.com/RIOT-OS/RIOT/pull/8822 is merged~~

cladmi commented:
The commits are normally all working and can be reviewed individually to see the whole process.

kaspar030 commented:
My gut feeling is that splitting this up would speed up reviewing.

kaspar030 commented:
> My gut feeling is that splitting this up would speed up reviewing.

Ah, seems like you're already doing that. ;)

miri64 commented:
Rebase required (unless there are more PRs this is based on)

cladmi commented:
As everything else has been merge, thanks @kaspar030, this PR now only includes the `flasher` changes and adding a `FLASHFILE` variable.

That is also a tricky change-set because it touches many flashers so could break stuff.

I will add "test" commits that prints output for debugging that it still works as before.

cladmi commented:
I am executing the following command `make --no-change-directory -C examples/hello-world flash-only` and checking the output difference with master with the following diff applied in a clean environment.

```
 flash: $(BUILD_BEFORE_FLASH) $(FLASHDEPS)
        $(call check_cmd,$(FLASHER),Flash program)
-       $(FLASHER) $(FFLAGS)
+       -@ timeout 5s $(FLASHER) $(FFLAGS)
 
 flash-only: flash
 
 preflash: $(BUILD_BEFORE_FLASH)
-       $(PREFLASHER) $(PREFFLAGS)
+       -$(PREFLASHER) $(PREFFLAGS)
```

The only difference I get between error messages is that `HEXFILE` variable has been renamed `BINFILE` in jlink.sh.

```
-HEXFILE found
-Error: Unable to locate HEXFILE
+BINFILE found
+Error: Unable to locate BINFILE
        (/home/harter/work/git/RIOT/examples/hello-world/bin/acd52832/hello-world.bin)
```

#### By printing the command

For a second run, I added an echo of `$(FLASHER) $(FFLAGS)` command, and I correctly get the same firmware argument.


* nz32_sc151 and spark-core only got quotes removed around filename (was already in FFLAGS)

* mount flash scripts with a fake mounted dir

cladmi commented:
### Output check

I am executing the following command make --no-change-directory -C examples/hello-world flash-only and checking the output difference with master with the following diff applied in a clean environment.

```
 flash: $(BUILD_BEFORE_FLASH) $(FLASHDEPS)
-       $(call check_cmd,$(FLASHER),Flash program)
+       -$(call check_cmd,$(FLASHER),Flash program)
-       $(FLASHER) $(FFLAGS)
+       -@ timeout 5s $(FLASHER) $(FFLAGS)
 
 flash-only: flash
 
 preflash: $(BUILD_BEFORE_FLASH)
-       $(PREFLASHER) $(PREFFLAGS)
+       -$(PREFLASHER) $(PREFFLAGS)
```
The only difference I get between error messages is that HEXFILE variable has been renamed BINFILE in jlink.sh.
```
-HEXFILE found
-Error: Unable to locate HEXFILE
+BINFILE found
+Error: Unable to locate BINFILE
        (/home/harter/work/git/RIOT/examples/hello-world/bin/acd52832/hello-world.bin)
```

#### With printing the command

For a second run, I added an echo of $(FLASHER) $(FFLAGS) command, and I correctly get the same firmware argument:
* Same command as before for boards that had it already
* Firmware argument matches the error message (openocd, jlink, copy flash.sh) where it was added
* Output is the same for boards,  `nz32_sc151` and `spark_core` where I removed the quotes around the firmware name, because I print it with  `echo`

cladmi commented:
Things that would be relevant to test before merging:

Test flashing with:

    * [ ] A board using openocd
    * [ ] Seeeduino_arch_pro using openocd with an `.hex` file
    * [ ] A board using jlink
    * [ ] `calliope-mini` or `mbed_lpc_178` or `microbit`
    * [ ] openocm904 that is using it's own flasher `robotis-loader.py`

cladmi commented:
Missing some boards here to test, in theory it should be OK but just verifying.

@basilfx Could you please try if it still flashes the `seeeduino_arch-pro` with this PR ?
@astralien3000 @vincent-d Could one of you two try flashing the `seeeduino_arch-pro` with this PR ?

Thank you :)

astralien3000 commented:
@cladmi I don't have any seeeduino, sorry.

cladmi commented:
@astralien3000 sorry was a copy paste error, I meant the `opencm904`, I saw you in the board history.

astralien3000 commented:
Ah, indeed, I can test. And I did, I was able to flash with this PR, no problem.

kaspar030 commented:
only seeeduino_arch-pro left?
I'd say, please squash!

cladmi commented:
I re-run the commands to check if a new board/flasher could have been added:
```
git grep FFLAGS | grep -v FLASHFILE
git grep OFLAGS |  grep -e '-O'
git grep HEXFILE | grep -v FLASHFILE
```

And it looked up to date.

cladmi commented:
I rebased on riot/master but I still need to update `boards/hifive1/dist/flasher.sh` and verify the rebased branch.

cladmi commented:
I split the commit flash in a separate PR to integrate it sooner.
I will split the new boards in upcoming ones.

kYc0o commented:
Please rebase, this can be in very soon.

emmanuelsearch commented:
@cladmi what is the way forward on this? @kaspar030 do you think we are close to merging this?

cladmi commented:
The flasher scripts are still required, and were split in https://github.com/RIOT-OS/RIOT/pull/9514.
Issue is that, I think, this PR was still fixing all flasher at the same time and got lost due to complexity of reviewing it at once. And is now outdated because of the new boards and flashers.

I should split it by flasher types to allow reviewing each part individually.
I am sometime missing `gerrit` feature where you can do `I approve and merge this single commit` and also thought it could be better to do all at once and forgot to split more to help.

cladmi commented:
I split the simplest ones in separate branches (== the ones that do not touch many boards).
I can do the other ones later.

cladmi commented:
There is still a lot of refactoring possible in the `esp` CPUs too. I should see what is the minimal required to use FLASHFILE in there and see if re-factoring should be done first or not.

aabadie commented:
#10539 is merged, so this one needs a rebase.

miri64 commented:
I asked @cladmi offline if it is ok to requeue this PR and he said yes, so I did (kind of want to get some PRs in today)

cladmi commented:
Rebased.

I am thinking about two way of going forward.

Right now, the `esp` cpus are not updated to use `FLASHFILE`. We can wait to update them first. It may also raise questions as they are, I think, somehow using multiple files for flashing.

Or introduce flashfile already for boards supporting it and migrate by batch.

@gschorcht maybe you have an idea how it should work for `esp` cpus.
With this I am planning to have a `FLASHFILE` variable that should be pointing to the file needed to flash. So you could say on another machine `FLASHFILE=file_received_by_some_way make flash-only`. However for some cases with the esp cpus, there are more than one file in practice.
Should I then move to a `FLASHFILES` and see how it could be handled or imagine that the other files will be created depending on `FLASHFILE` ? Regarding the binaries that are in the repo, we can consider them as static/toolchain file so not part of this.

gschorcht commented:
@cladmi I will take a look, when I'm back in my office on Monday.

cladmi commented:
@gschorcht thanks :)

gschorcht commented:
@cladmi 
For ESP8266 boards, `esptool.py` is used to generate two binary files from the `$(ELFFILE)` file

`$(ELFFILE)-0x00000.bin` that has to be flashed to address `0x00000` and 
`$(ELFFILE)-0x10000.bin` that has to be flashed to address `0x10000`.

The file names of these binaries are derived from input file name  `$(ELFFILE)` by the `esptool.py` and are fix, the ``$(ELFFILE)-<address>.bin`. That is, there are always two files that have to be flashed. The only single file is the `$(ELFFILE)` file. Therefore, `$(ELFFILE)` is the only file that might be used for `FLASHFILE` variable.

gschorcht commented:
@cladmi For ESP32 boards it is a bit different. ESP32 boards require to flash three files:
- the bootloader has to be flashed to address `0x1000`
- the partition table has to be flashed  to address `0x8000`
- the image `$(ELFFILE).bin` has to be flashed to address `0x10000`

While the bootloader is a static binary file and has not to be generated, the partition table is generated from the image file `$(ELFFILE).bin` to use correct size. The generation of the `$(ELFFILE).bin` file should be part of the make, while the generation of the partition table could be part of flashing. 

So, if I understand your changes correctly, it should be possible to use `FLASHFILE` variable if the output of `esptool.py` is written to `$(BINFILE)` instead of `$(ELFFILE).bin`. I can try to change it in that way.

I would take the chance also to move flash related parts of `cpu/esp32/Makeile.include` to `boards/common/esp32/Makefile.include`. Or should it then be done in another PR? I didn't know that flash related parts have to be defined in board makefiles when I started the ESP32 port. Because it is exactly the same for all ESP32 boards, I thought it can be part of the MCU makefile.

kaspar030 commented:
> I would take the chance also to move flash related parts of `cpu/esp32/Makeile.include` to `boards/common/esp32/Makefile.include`. Or should it then be done in another PR? I didn't know that flash related parts have to be defined in board makefiles when I started the ESP32 port. Because it is exactly the same for all ESP32 boards, I thought it can be part of the MCU makefile.

If it is exactly the same for all ESP32 boards, it can stay in cpu/esp32.

gschorcht commented:
@cladmi Just to be sure how it should be before I change anything. What should be the output of make `all`, only the ELF file or already all the binaries required for flashing?

At the moment, the target `all` for ESP boards only generates the ELF file. The target `flash` or` flash-only` depends on `FLASHDEPS`, which contains the target `preflash` for ESP boards. The target `preflash` generates all necessary binaries for flashing.

I'm asking because the target `preflash` contains UNIX shell commands. Creating an image in Docker and flashing on Windows with `esptool.py` is currently not possible. If all binary files would already be created by `make all`, flashing with` esptool.py` would also be possible on Windows. Alternatively, the user can of course execute `make preflash` in the Docker and on Windows only `esptool.py`. But then I should document that.

At this point I'm wondering whether the possibility to flash on Windows is desired at all? I have seen some documentation in the Wiki, but only for ARM platforms if I'm right.

gschorcht commented:
> In my idea, I would like it to be a file, that a user could flash without RIOT. A file where a user could flash with the gui he received from his board manufacturer. A file you could download from a website to have a nice esp32 firmware doing monitoring without knowing it is running RIOT. So the file would be all the information required to flash and being in a "standard" format to flash with out of the box tool.

Pretty cool vision. In that case you wouldn't have the RIOT make build system at all. For the ESPs this would mean that only `esptool.py` has to be installed.

For the ESP8266, a single `FLASHFILE` would be enough, since the two files that are flashed with `esptool.py` in the second step could be generated in a first step with the `esptool.py` itself, although this is not trivial.

For ESP32, however, the user would need all three files `bootloader.bin`,` partitiontable.bin` and `application.bin` as` FLASHFILES`, which need to be flashed to specific addresses.

> In a more developer oriented goal, the idea is that you could send this file to somebody, and he could do `FLASHFILE=this_local_file BOARD=myboard make flash-only` like we do in the CI where the flashing machine is different from the building one.

In that case, the RIOT make build system is available and it is indeed enough to set `FLASHFILE` to `$(ELFFILE)` and replace all occurrences of `$(ELFFILE)` by `$(FLASHFILE)` in `cpu/esp8266/Makefile.include` and `cpu/esp32/Makefile.include`

```diff
--- a/cpu/esp32/Makefile.include
+++ b/cpu/esp32/Makefile.include
@@ -141,12 +141,12 @@ FLASH_SIZE ?= 16m
 export PREFLASHER = $(ESPTOOL)
 export PREFFLAGS  = --chip esp32 elf2image
 export PREFFLAGS += -fm $(FLASH_MODE) -fs $(FLASH_SIZE) -ff $(FLASH_FREQ)
-export PREFFLAGS += -o $(ELFFILE).bin $(ELFFILE);
+export PREFFLAGS += -o $(FLASHFILE).bin $(FLASHFILE);
 export PREFFLAGS += echo "" > $(BINDIR)/partitions.csv;
 export PREFFLAGS += echo "nvs, data, nvs, 0x9000, 0x6000" >> $(BINDIR)/partitions.csv;
 export PREFFLAGS += echo "phy_init, data, phy, 0xf000, 0x1000" >> $(BINDIR)/partitions.csv;
 export PREFFLAGS += echo -n "factory, app, factory, 0x10000, " >> $(BINDIR)/partitions.csv;
-export PREFFLAGS += ls -l $(ELFFILE).bin | awk '{ print $$5 }' >> $(BINDIR)/partitions.csv;
+export PREFFLAGS += ls -l $(FLASHFILE).bin | awk '{ print $$5 }' >> $(BINDIR)/partitions.csv;
 
 export PREFFLAGS += python $(RIOTCPU)/$(CPU)/gen_esp32part.py --disable-sha256sum
 export PREFFLAGS += --verify $(BINDIR)/partitions.csv $(BINDIR)/partitions.bin
@@ -161,7 +161,7 @@ ifeq ($(QEMU), 1)
     export FFLAGS += head -c $$((0x8000)) |
     export FFLAGS += cat - $(BINDIR)/partitions.bin tmp.bin |
     export FFLAGS += head -c $$((0x10000)) |
-    export FFLAGS += cat - $(ELFFILE).bin tmp.bin |
+    export FFLAGS += cat - $(FLASHFILE).bin tmp.bin |
     export FFLAGS += head -c $$((0x400000)) > $(BINDIR)/esp32flash.bin && rm tmp.bin &&
     export FFLAGS += cp $(RIOTCPU)/$(CPU)/bin/rom_0x3ff90000_0x00010000.bin $(BINDIR)/rom1.bin &&
     export FFLAGS += cp $(RIOTCPU)/$(CPU)/bin/rom_0x40000000_0x000c2000.bin $(BINDIR)/rom.bin
@@ -173,5 +173,5 @@ else
     export FFLAGS += -z -fm $(FLASH_MODE) -fs detect -ff $(FLASH_FREQ)
     export FFLAGS += 0x1000 $(RIOTCPU)/$(CPU)/bin/bootloader.bin
     export FFLAGS += 0x8000 $(BINDIR)/partitions.bin
-    export FFLAGS += 0x10000 $(ELFFILE).bin
+    export FFLAGS += 0x10000 $(FLASHFILE).bin
 endif
```
```diff
--- a/cpu/esp8266/Makefile.include
+++ b/cpu/esp8266/Makefile.include
@@ -127,21 +127,21 @@ USEMODULE += xtensa
 # configure preflasher to convert .elf to .bin before flashing
 FLASH_SIZE = -fs 8m
 export PREFLASHER ?= esptool.py
-export PREFFLAGS  ?= elf2image $(FLASH_SIZE) $(ELFFILE)
+export PREFFLAGS  ?= elf2image $(FLASH_SIZE) $(FLASHFILE)
 export FLASHDEPS  ?= preflash
 
 # flasher configuration
 ifeq ($(QEMU), 1)
     export FLASHER = cat
-    export FFLAGS += $(ELFFILE)-0x00000.bin /dev/zero | head -c $$((0x10000)) | cat -
-    export FFLAGS += $(ELFFILE)-0x10000.bin /dev/zero | head -c $$((0xfc000)) | cat -
-    export FFLAGS += $(RIOTCPU)/$(CPU)/bin/esp_init_data_default.bin > $(ELFFILE).bin
+    export FFLAGS += $(FLASHFILE)-0x00000.bin /dev/zero | head -c $$((0x10000)) | cat -
+    export FFLAGS += $(FLASHFILE)-0x10000.bin /dev/zero | head -c $$((0xfc000)) | cat -
+    export FFLAGS += $(RIOTCPU)/$(CPU)/bin/esp_init_data_default.bin > $(FLASHFILE).bin
 else
     FLASH_MODE ?= dout
     export PROGRAMMER_SPEED ?= 460800
     export FLASHER = esptool.py
     export FFLAGS += -p $(PORT) -b $(PROGRAMMER_SPEED) write_flash
     export FFLAGS += -fm $(FLASH_MODE)
-    export FFLAGS += 0 $(ELFFILE)-0x00000.bin
-    export FFLAGS += 0x10000 $(ELFFILE)-0x10000.bin
+    export FFLAGS += 0 $(FLASHFILE)-0x00000.bin
+    export FFLAGS += 0x10000 $(FLASHFILE)-0x10000.bin
 endif
```

I have tested it and it works as expected. Probably, the easiest way that you change it in your PR.

cladmi commented:
This PR now still has a big first commit doing everything as the same time as forcing the BS API change.
I will split with a first commit introducing `FLASHFILE` and migration by families, would simplify reviewing and have the enforcing at the end.

cladmi commented:
I rebased and kept the `export FFLAGS` as they were before.
Some commits change for more than one programmer.

I should split the first commits to help starting the transition as FLASHFILE would be in place.

This showed me that their are several flashers where there is no common `tools/flasher` for it so a lot is duplicated. I think it may be a good idea to migrate them at the same time.

I could suggest that, for each flasher class:

* create a `makefiles/tools/flasher/flasher_name.inc.mk` (and so migrate openocd to it for example with tmp migrations warnings)
* remove all useless `exports`
* Put config in common with namespaced variable instead of just putting `FFLAGS += -p atmegablah`
* Use `FLASHFILE` in the common one.
* If an export should be done for a script, it must be exported in the common file ex `OPENOCD_ADAPTER_INIT`

What do you think ? Or should it be done as a separate task ?
There would be a dedicated tracking issue.

Removing FFLAGS exports is something needed for https://github.com/RIOT-OS/RIOT/pull/10440 so could be good to do it all together.

cladmi commented:
As there was not feedback on the long version.
I will move forward for this one by introducing the variable first unused.
Then migrate boards by batch.
This would allow moving forward and using it for parts that already need the `flashfile` variable.

I often forget that it is easier to separate in commits that touch a limited set of boards even if it requires a transition step.

cladmi commented:
Re-run check to see what has not been updated recently.

cladmi commented:
This pull request is now not waiting for other pull requests anymore and I added the end of migration commit.

The test would prevent compilation:

```
FLASHFILE= make --no-print-directory -C examples/hello-world/
/home/harter/work/git/RIOT/examples/hello-world/../../Makefile.include:447: *** FLASHFILE is not defined for this board: .  Stop.
```

In the current form, it would ignore compiling it in `murdock` as `info-toolchains-supported` would return nothing (and not fail).
I could change it to not error when `info-toolchains-supported` is used if requested, but not sure if it makes it more maintainable.

```
FLASHFILE= make --no-print-directory -C examples/hello-world/ info-toolchains-supported
/home/harter/work/git/RIOT/examples/hello-world/../../Makefile.include:447: *** FLASHFILE is not defined for this board: .  Stop.
```

-----

I re-did the "Verifying after rebase" and found one remaining instance of setting `OFLAGS` format. It is split in #11849 but not part of this one as it is now just a duplication cleanup.

tcschmidt commented:
@jia200x @leandrolanzieri how about this? Can this move?

cladmi commented:
I rebased to master to simplify testing in case new boards would be broken.

And currently all boards correctly define FLASHFILE as this goes to the end:

```
for board in $(make info-boards); do BOARD=${board} make --no-print-directory -C examples/hello-world/ info-debug-variable-FLASHFILE || break; done 2>&1 | sed "s|${PWD}/||" | tee output
```

<details><summary>No error on any boards</summary>

```
examples/hello-world/bin/acd52832/hello-world.bin
examples/hello-world/bin/airfy-beacon/hello-world.elf
examples/hello-world/bin/arduino-due/hello-world.bin
examples/hello-world/bin/arduino-duemilanove/hello-world.hex
examples/hello-world/bin/arduino-leonardo/hello-world.hex
examples/hello-world/bin/arduino-mega2560/hello-world.hex
examples/hello-world/bin/arduino-mkr1000/hello-world.bin
examples/hello-world/bin/arduino-mkrfox1200/hello-world.bin
examples/hello-world/bin/arduino-mkrzero/hello-world.bin
examples/hello-world/bin/arduino-nano/hello-world.hex
examples/hello-world/bin/arduino-uno/hello-world.hex
examples/hello-world/bin/arduino-zero/hello-world.bin
examples/hello-world/bin/avsextrem/hello-world.hex
examples/hello-world/bin/b-l072z-lrwan1/hello-world.elf
examples/hello-world/bin/b-l475e-iot01a/hello-world.elf
examples/hello-world/bin/blackpill/hello-world.elf
examples/hello-world/bin/bluepill/hello-world.elf
examples/hello-world/bin/calliope-mini/hello-world.elf
Warning: no PORT set!
examples/hello-world/bin/cc2538dk/hello-world.bin
examples/hello-world/bin/cc2650-launchpad/hello-world.elf
examples/hello-world/bin/cc2650stk/hello-world.elf
examples/hello-world/bin/chronos/hello-world.hex
examples/hello-world/bin/ek-lm4f120xl/hello-world.elf
ESP32_SDK_DIR should be defined as /path/to/esp-idf directory
ESP32_SDK_DIR is set by default to /opt/esp/esp-idf
examples/hello-world/bin/esp32-mh-et-live-minikit/hello-world.elf
ESP32_SDK_DIR should be defined as /path/to/esp-idf directory
ESP32_SDK_DIR is set by default to /opt/esp/esp-idf
examples/hello-world/bin/esp32-olimex-evb/hello-world.elf
ESP32_SDK_DIR should be defined as /path/to/esp-idf directory
ESP32_SDK_DIR is set by default to /opt/esp/esp-idf
examples/hello-world/bin/esp32-wemos-lolin-d32-pro/hello-world.elf
ESP32_SDK_DIR should be defined as /path/to/esp-idf directory
ESP32_SDK_DIR is set by default to /opt/esp/esp-idf
examples/hello-world/bin/esp32-wroom-32/hello-world.elf
ESP32_SDK_DIR should be defined as /path/to/esp-idf directory
ESP32_SDK_DIR is set by default to /opt/esp/esp-idf
examples/hello-world/bin/esp32-wrover-kit/hello-world.elf
ESP8266_NEWLIB_DIR should be defined as /path/to/newlib directory
ESP8266_NEWLIB_DIR is set by default to /opt/esp/newlib-xtensa
ESP8266_SDK_DIR should be defined as /path/to/sdk directory
ESP8266_SDK_DIR is set by default to /opt/esp/esp-open-sdk/sdk
examples/hello-world/bin/esp8266-esp-12x/hello-world.elf
ESP8266_NEWLIB_DIR should be defined as /path/to/newlib directory
ESP8266_NEWLIB_DIR is set by default to /opt/esp/newlib-xtensa
ESP8266_SDK_DIR should be defined as /path/to/sdk directory
ESP8266_SDK_DIR is set by default to /opt/esp/esp-open-sdk/sdk
examples/hello-world/bin/esp8266-olimex-mod/hello-world.elf
ESP8266_NEWLIB_DIR should be defined as /path/to/newlib directory
ESP8266_NEWLIB_DIR is set by default to /opt/esp/newlib-xtensa
ESP8266_SDK_DIR should be defined as /path/to/sdk directory
ESP8266_SDK_DIR is set by default to /opt/esp/esp-open-sdk/sdk
examples/hello-world/bin/esp8266-sparkfun-thing/hello-world.elf
examples/hello-world/bin/f4vi1/hello-world.bin
examples/hello-world/bin/feather-m0/hello-world.bin
examples/hello-world/bin/firefly/hello-world.bin
examples/hello-world/bin/fox/hello-world.elf
examples/hello-world/bin/frdm-k22f/hello-world.elf
examples/hello-world/bin/frdm-k64f/hello-world.elf
examples/hello-world/bin/frdm-kw41z/hello-world.elf
examples/hello-world/bin/hamilton/hello-world.bin
examples/hello-world/bin/hifive1/hello-world.elf
examples/hello-world/bin/hifive1b/hello-world.bin
examples/hello-world/bin/i-nucleo-lrwan1/hello-world.elf
examples/hello-world/bin/ikea-tradfri/hello-world.bin
examples/hello-world/bin/iotlab-a8-m3/hello-world.elf
examples/hello-world/bin/iotlab-m3/hello-world.elf
examples/hello-world/bin/jiminy-mega256rfr2/hello-world.hex
examples/hello-world/bin/limifrog-v1/hello-world.elf
examples/hello-world/bin/lobaro-lorabox/hello-world.bin
examples/hello-world/bin/lsn50/hello-world.elf
examples/hello-world/bin/maple-mini/hello-world.elf
examples/hello-world/bin/mbed_lpc1768/hello-world.bin
examples/hello-world/bin/mega-xplained/hello-world.hex
examples/hello-world/bin/microbit/hello-world.elf
examples/hello-world/bin/msb-430/hello-world.hex
examples/hello-world/bin/msb-430h/hello-world.hex
examples/hello-world/bin/msba2/hello-world.hex
examples/hello-world/bin/msbiot/hello-world.elf
examples/hello-world/bin/mulle/hello-world.elf
examples/hello-world/bin/native/hello-world.elf
examples/hello-world/bin/nrf51dk/hello-world.elf
examples/hello-world/bin/nrf51dongle/hello-world.bin
examples/hello-world/bin/nrf52832-mdk/hello-world.hex
examples/hello-world/bin/nrf52840-mdk/hello-world.hex
examples/hello-world/bin/nrf52840dk/hello-world.bin
examples/hello-world/bin/nrf52dk/hello-world.bin
examples/hello-world/bin/nrf6310/hello-world.bin
examples/hello-world/bin/nucleo-f030r8/hello-world.elf
examples/hello-world/bin/nucleo-f031k6/hello-world.elf
examples/hello-world/bin/nucleo-f042k6/hello-world.elf
examples/hello-world/bin/nucleo-f070rb/hello-world.elf
examples/hello-world/bin/nucleo-f072rb/hello-world.elf
examples/hello-world/bin/nucleo-f091rc/hello-world.elf
examples/hello-world/bin/nucleo-f103rb/hello-world.elf
examples/hello-world/bin/nucleo-f207zg/hello-world.elf
examples/hello-world/bin/nucleo-f302r8/hello-world.elf
examples/hello-world/bin/nucleo-f303k8/hello-world.elf
examples/hello-world/bin/nucleo-f303re/hello-world.elf
examples/hello-world/bin/nucleo-f303ze/hello-world.elf
examples/hello-world/bin/nucleo-f334r8/hello-world.elf
examples/hello-world/bin/nucleo-f401re/hello-world.elf
examples/hello-world/bin/nucleo-f410rb/hello-world.elf
examples/hello-world/bin/nucleo-f411re/hello-world.elf
examples/hello-world/bin/nucleo-f412zg/hello-world.elf
examples/hello-world/bin/nucleo-f413zh/hello-world.elf
examples/hello-world/bin/nucleo-f429zi/hello-world.elf
examples/hello-world/bin/nucleo-f446re/hello-world.elf
examples/hello-world/bin/nucleo-f446ze/hello-world.elf
examples/hello-world/bin/nucleo-f722ze/hello-world.elf
examples/hello-world/bin/nucleo-f746zg/hello-world.elf
examples/hello-world/bin/nucleo-f767zi/hello-world.elf
examples/hello-world/bin/nucleo-l031k6/hello-world.elf
examples/hello-world/bin/nucleo-l053r8/hello-world.elf
examples/hello-world/bin/nucleo-l073rz/hello-world.elf
examples/hello-world/bin/nucleo-l152re/hello-world.elf
examples/hello-world/bin/nucleo-l432kc/hello-world.elf
examples/hello-world/bin/nucleo-l433rc/hello-world.elf
examples/hello-world/bin/nucleo-l452re/hello-world.elf
examples/hello-world/bin/nucleo-l476rg/hello-world.elf
examples/hello-world/bin/nucleo-l496zg/hello-world.elf
examples/hello-world/bin/nucleo-l4r5zi/hello-world.elf
examples/hello-world/bin/nz32-sc151/hello-world.bin
examples/hello-world/bin/opencm904/hello-world.bin
examples/hello-world/bin/openmote-b/hello-world.hex
examples/hello-world/bin/openmote-cc2538/hello-world.bin
examples/hello-world/bin/particle-argon/hello-world.hex
examples/hello-world/bin/particle-boron/hello-world.hex
examples/hello-world/bin/particle-xenon/hello-world.hex
examples/hello-world/bin/pba-d-01-kw2x/hello-world.elf
examples/hello-world/bin/phynode-kw41z/hello-world.elf
examples/hello-world/bin/pic32-clicker/hello-world.hex
examples/hello-world/bin/pic32-wifire/hello-world.hex
examples/hello-world/bin/pyboard/hello-world.bin
examples/hello-world/bin/reel/hello-world.elf
examples/hello-world/bin/remote-pa/hello-world.bin
examples/hello-world/bin/remote-reva/hello-world.bin
examples/hello-world/bin/remote-revb/hello-world.bin
examples/hello-world/bin/ruuvitag/hello-world.bin
examples/hello-world/bin/samd21-xpro/hello-world.bin
examples/hello-world/bin/same54-xpro/hello-world.bin
examples/hello-world/bin/saml10-xpro/hello-world.bin
examples/hello-world/bin/saml11-xpro/hello-world.bin
examples/hello-world/bin/saml21-xpro/hello-world.bin
examples/hello-world/bin/samr21-xpro/hello-world.bin
examples/hello-world/bin/samr30-xpro/hello-world.bin
examples/hello-world/bin/seeeduino_arch-pro/hello-world.hex
examples/hello-world/bin/sensebox_samd21/hello-world.bin
examples/hello-world/bin/slstk3401a/hello-world.bin
examples/hello-world/bin/slstk3402a/hello-world.bin
examples/hello-world/bin/sltb001a/hello-world.bin
examples/hello-world/bin/slwstk6000b/hello-world.bin
examples/hello-world/bin/slwstk6220a/hello-world.bin
examples/hello-world/bin/sodaq-autonomo/hello-world.bin
examples/hello-world/bin/sodaq-explorer/hello-world.bin
examples/hello-world/bin/sodaq-one/hello-world.bin
examples/hello-world/bin/sodaq-sara-aff/hello-world.bin
examples/hello-world/bin/spark-core/hello-world.bin
examples/hello-world/bin/stk3600/hello-world.bin
examples/hello-world/bin/stk3700/hello-world.bin
examples/hello-world/bin/stm32f0discovery/hello-world.elf
examples/hello-world/bin/stm32f3discovery/hello-world.elf
examples/hello-world/bin/stm32f429i-disc1/hello-world.elf
examples/hello-world/bin/stm32f4discovery/hello-world.elf
examples/hello-world/bin/stm32f769i-disco/hello-world.elf
examples/hello-world/bin/stm32l0538-disco/hello-world.elf
examples/hello-world/bin/stm32l476g-disco/hello-world.elf
examples/hello-world/bin/teensy31/hello-world.hex
examples/hello-world/bin/telosb/hello-world.hex
examples/hello-world/bin/thingy52/hello-world.bin
examples/hello-world/bin/ublox-c030-u201/hello-world.elf
examples/hello-world/bin/udoo/hello-world.bin
examples/hello-world/bin/usb-kw41z/hello-world.elf
examples/hello-world/bin/waspmote-pro/hello-world.hex
examples/hello-world/bin/wsn430-v1_3b/hello-world.hex
examples/hello-world/bin/wsn430-v1_4/hello-world.hex
examples/hello-world/bin/yunjia-nrf51822/hello-world.elf
examples/hello-world/bin/z1/hello-world.hex
```
</details>

cladmi commented:
I think it would not complain for the boards in 
`TEST_BOARDS_LLVM_COMPILE:="iotlab-m3 native nrf52dk mulle nucleo-f401re samr21-xpro slstk3402a"` if `FLASHFILE` is not defined as it would be removed by

https://github.com/RIOT-OS/RIOT/blob/552e425b66373f1eb0948ca4f0c93a698646db7b/.murdock#L120-L124

But the test from https://github.com/RIOT-OS/RIOT/pull/8838#issuecomment-522603844 would have reported an issue.
And any local compilation would fail.

cladmi commented:
As I was asked precision about my comment IRL.

For the boards in `TEST_BOARDS_LLVM_COMPILE`, if the `FLASHFILE` was not defined, the reported supported toolchains would be empty, and I think it would be ignored in `murdock` compilation.

```
FLASHFILE= BOARD=iotlab-m3 make -s --no-print-directory -C examples/hello-world/  info-toolchains-supported 2> /dev/null | grep -o -e "llvm" -e "gnu"
# empty answer
```

However, the test in my https://github.com/RIOT-OS/RIOT/pull/8838#issuecomment-522603844 comment would have reported error if a current board was not setting `FLASHFILE` with the default `PROGRAMMER` configuration.

If a new board is added, or boards are refactored, a local compilation should also be done which would report the missing `FLASHFILE`.

I could add an handling for this that detects if the reported toolchain is empty. However, for non default `PROGRAMMER`, we would would also only see it when trying to compile for them as it will not be checked by murdock.

Also, any other errors during the makefile (that is not info-boards-supported) would make the board ignored by murdock so it is not specific to this.

MrKevinWeiss commented:
Thanks update.  I still think this is good to go!

cladmi commented:
Thank you for the review.

Took 1 year and a half to get rid of such a bad pattern D: Finally done.
